CONTRIBUTING.md

```diff
@@ -31,10 +31,135 @@ Please be aware of the following things when filing bug reports.
+
+The next interesting thing is `self._state`. The library uses a state-centric design, which passes the state around to most objects. This is what enables edit methods and most methods on objects. Every Discord model that makes requests uses that internal state and its http attribute to make requests to the Discord API. Each endpoint is processed and defined in [disnake/http.py](disnake/http.py). This is where all of the methods are defined--and its where `pins_from` is defined too!
+
+The source of `pins_from` is as follows:
+
+```py
+    def pins_from(self, channel_id: Snowflake) -> Response[List[message.Message]]:
+        return self.request(Route("GET", "/channels/{channel_id}/pins", channel_id=channel_id))
+```
+
+This is the basic model that all API request methods follow. Define the `Route`, provide the major parameters (in this example `channel_id`), then return a call to self.request(), and finally typehints.
+
+The `Response[]` part in this typehint is referring to `self.request`, as the important thing here is that `pins_from` is **not** a coroutine. Rather, pins_from does preprocessing and self.request does the actual work. The result from `pins_from` is awaited by `disnake.abc.Messageable.pins`.
+
+The Route class is how all routes are processed internally. This, along with `self.request` makes it possible to properly handle all ratelimits, and rarely encounter 429s. This is additionally why `channel_id` is provided as a kwarg to `Route`.
+
+### Writing Documentation
+
+While a new feature can be useful, it requires documentation to be useful too. When updating a class or method, we ask that you use
+[sphinx directives](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded) in the docstring to notate when it was updated, and what about it was updated.
+
+For example, the docstring below for `pins()`.
+
+```py
+      """|coro|
+
+      Retrieves all messages that are currently pinned in the channel.
+
+      .. note::
+
+          Due to a limitation with the Discord API, the :class:`.Message`
+          objects returned by this method do not contain complete
+          :attr:`.Message.reactions` data.
+
+      Raises
+      ------
+      HTTPException
+          Retrieving the pinned messages failed.
+
+      Returns
+      -------
+      List[:class:`.Message`]
+          The messages that are currently pinned.
+      """
+```
+
+If we were to add a new parameter to this method, we would need to add a few things to this docstring. Lets pretend we're adding a parameter, ``oldest_first``.
+
+We use NumPy style docstrings parsed with sphinx's napoleon extension, the primary documentation for these docstrings is [here](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html).
```

Comment:
Don't think linking to the napoleon docs is a good idea, since the format outlined there doesn't quite match the standards used in the library, particularly the ` : ` spacing and the lack of `:class:` directives

Reply:
Most of the format there *is* the same, so I think its worth it. At the very least, I learned stuff from it when I looked at it to add to these docs.